usaspending_api/awards/v2/data_layer/orm.py

```diff
@@ -577,3 +585,91 @@ def fetch_naics_hierarchy(naics: str) -> dict:
+        from financial_accounts_by_awards faba
+        inner join submission_attributes s on faba.submission_id=s.submission_id
+        where {award_id_sql} and faba.disaster_emergency_fund_code is not null
+    ),
+    closed_periods as (
+        -- This is NOT right, since period dates are not the same as submission window close dates,
+        -- but an approximation for illustration purposes until we have the reporting dates table up
+        select
+            distinct concat(p.reporting_fiscal_year::text,
+            lpad(p.reporting_fiscal_period::text, 2, '0')) as fyp,
+            p.reporting_fiscal_year, p.reporting_fiscal_period
+        from submission_attributes p
+        where now()::date > p.reporting_period_end -- change "period end" with "window close"
+        order by p.reporting_fiscal_year desc, p.reporting_fiscal_period desc
+    ),
+    fy_final_balances as (
+        -- Rule: If a balance is not zero at the end of the year, it must be reported in the
+        -- final period's submission (month or quarter), otherwise assume it to be zero
+        select faba.award_id, sum(faba.gross_outlay_amount_by_award_cpe) as prior_fys_outlay,
+            faba.disaster_emergency_fund_code
+        from eligible_file_c_records faba
+        group by
+            faba.award_id,
+            faba.reporting_fiscal_period,
+            faba.disaster_emergency_fund_code
+        having faba.reporting_fiscal_period = 12
+        and sum(faba.gross_outlay_amount_by_award_cpe) > 0
+    ),
+    current_fy_balance as (
+        select
+            faba.award_id,
+            faba.reporting_fiscal_year,
+            faba.reporting_fiscal_period,
+            faba.disaster_emergency_fund_code,
+            sum(faba.gross_outlay_amount_by_award_cpe) as current_fy_outlay
+        from eligible_file_c_records faba
+        group by
+            faba.award_id,
+            faba.reporting_fiscal_year,
+            faba.reporting_fiscal_period,
+            faba.disaster_emergency_fund_code
+        having concat(faba.reporting_fiscal_year::text, lpad(faba.reporting_fiscal_period::text, 2, '0')) in
+        (select max(fyp) from closed_periods) and sum(faba.gross_outlay_amount_by_award_cpe) > 0)
+    select
+        faba.disaster_emergency_fund_code,
+        coalesce(ffy.prior_fys_outlay, 0) + coalesce(cfy.current_fy_outlay, 0) as total_outlay,
+        coalesce(sum(faba.transaction_obligated_amount), 0) as obligated_amount
+    from eligible_file_c_records faba
+    left join fy_final_balances ffy on ffy.award_id = faba.award_id and ffy.disaster_emergency_fund_code = faba.disaster_emergency_fund_code
+    left join current_fy_balance cfy
+        on cfy.reporting_fiscal_period != 12 -- don't duplicate the year-end period's value if in unclosed period 01
+        and cfy.award_id = faba.award_id
+        and cfy.disaster_emergency_fund_code = faba.disaster_emergency_fund_code
+    group by faba.disaster_emergency_fund_code, total_outlay;
```

Comment:
I was having a hard time following this all the way through, but tested and it verifies the group by to be okay. But to confirm, you can group-by `total_outlay` here, because effectively this query is going to _repeat_ the same outlay+defc values for each distinct obligation amount present for the award in the `faba` table? True?

This is a nice way to get outlay and obligation sums in 1 query (if a bit hard to follow) in order to not have to run 2 similar queries for each of outlay and obligation.

Reply:
Yes, the outlays are calculated in the CTEs (current fy balance and fy final balances), so they are the same for every entry in the faba CTE. They get grouped by because they aren't aggregated like obligations.